sound/soc/sof/ipc4-pcm.c

```diff
@@ -32,6 +34,174 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
 }
 EXPORT_SYMBOL(sof_ipc4_set_pipeline_state);
 
+static u64 ipc4_cc_read(const struct cyclecounter *cc)
+{
+	struct snd_sof_pcm_stream *spcm = container_of(cc, struct snd_sof_pcm_stream, cc);
+
+	return spcm->llp_slot->reading.wclk_l + ((u64)spcm->llp_slot->reading.wclk_u << 32) -
+			spcm->stream_start_wclk;
+}
+
+static void sof_ipc4_timecounter_init(struct snd_sof_dev *sdev, struct snd_sof_pcm_stream *spcm,
+				      u32 wclk)
+{
+	struct cyclecounter *cc = &spcm->cc;
+	u32 maxsec;
+
+	cc->read = ipc4_cc_read;
+	cc->mask = CLOCKSOURCE_MASK(32);
+
+	/* The counter wraps around after 2^32 -1 / wclk seconds. */
+	maxsec = BIT(32) - 1;
+	do_div(maxsec, wclk);
+	clocks_calc_mult_shift(&cc->mult, &cc->shift, wclk, NSEC_PER_SEC, maxsec);
+
+	/* convert 1 cycle to X ns */
+	spcm->accuracy = (1ULL * cc->mult) >> cc->shift;
+	dev_dbg(sdev->dev, "accuracy %d\n", spcm->accuracy);
+}
+
+/*
+ * calculate stream start time and resolution of wclk in dsp
+ * The llp position in memory window is not the real position for a stream in two cases
+ * (1) dai starts before host in multiple pipeline cases. The llp conunter works before
+ *  stream is started.
+ * (2) multiple streams with mixer. The llp counter is same for all streams.
+ *
+ * the solution is using stream_start_offset for each stream to save the llp counter
+ * when that stream is started. stream_start_offset is based on sample size, so need
+ * to convert it to sample count.
+ * the stream position = current llp value  - stream_start_offset
+ *
+ * to calculate stream start time (wclk0), we use math interpolation
+ *	wclk0				wclk1					wclk2
+ * stream_start_offsets*		llp1					llp2
+ *
+ * wclk0 = wclk2 - (wclk1 -wclk2)/(llp1 -llp2) * (llp2  - stream_start_offsets)
```

Comment:
No wclk0 saved.   
(1) The division is only built once for start time and not calculated during stream running again, so no long term drift. And it is built with u64 not u32 since llp position and wclk are all u64
(2) we can also build start time with wclk1 - (llp1 - stream_start_offsets) / sampling_frequency, but it is not precision as above

Reply:
Maybe I need more coffee but what is wclk0 used for later? Since it's an approximated value, all future results based on wclk0 will be tainted by some level of fixed-point arithmetic.

Just trying to gauge how this impacts long-term results.

It would help if you had all the steps described in a comment, with the variables in the comment and the code matching. I don't see wclk0 and wclk1 being used below.

Reply:
I update the variable name to follow the comments.  wclk0 should be the stream start clk in dma.
It only affect the start time and calculation is done in 64bit so the error should be the last bit, about 24 nanosecond for 24MHZ.